tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+        rows = (pd.Series(selfrows, index=selfrows)
+                .loc[new_bbox[3]:new_bbox[1]].values)
+        cols = (pd.Series(selfcols, index=selfcols)
+                .loc[new_bbox[0]:new_bbox[2]].values)
+
+        # clip existing grids to new bbox
+        for i in self.gridlist:
+            data = pd.DataFrame(getattr(self, i),
+                                index=selfrows, columns=selfcols)
+            data = (data.reindex(rows).reindex_axis(cols, axis=1)
+                        .fillna(self.nodata[i]).values)
+            setattr(self, i, data)
+        
+        # set class attributes
+        self.bbox = tuple(new_bbox)
+        self.shape = tuple([len(rows), len(cols)])
+
+    def set_nodata(self, data_name, new_nodata, old_nodata=None):
+        """
+        Change nodata value of dataset.
+
+        Parameters
+        ----------
+        data_name : string
+                    Attribute name of dataset to change
+        new_nodata : int or float
+                     New nodata value to use
+        old_nodata : int or float (optional)
+                     If none provided, defaults to self.nodata[data_name]
+        """
+
+        if old_nodata is None:
+            old_nodata = self.nodata[data_name]
+        data = getattr(self, data_name)
+        np.place(data, data==old_nodata, new_nodata)
+        self.nodata.update({data_name : new_nodata})
+
+    def catchment_mask(self, to_mask, mask_source='catch'):
+        """
+        Masks grid cells not included in catchment.
+
+        Parameters
+        ----------
+        to_mask : string
+                  name of dataset to mask
+        mask_source : string (optional)
+                      dataset on which mask is based (defaults to 'catch')
+        """
+
+        mask = (getattr(self, mask_source) == self.nodata[mask_source])
+        if isinstance(to_mask, str):
+            np.place(getattr(self, to_mask), mask, self.nodata[to_mask])
+        elif isinstance(to_mask, (list, tuple, np.ndarray)):
+            for i in to_mask:
+                np.place(getattr(self, i), mask, self.nodata[i])
+
+    def to_ascii(self, data_name=None, file_name=None, delimiter=' ', **kwargs):
+        """
+        Writes grid data to ascii grid files.
+
+        Parameters
+        ----------
+        data_name : string or list-like (optional)
+                    Attribute name(s) of datasets to write. Defaults to all
+                    grid dataset names.
+        file_name : string or list-like
+                    Name(s) of file(s) to write to
+        delimiter : string
+                    Delimiter to use
+        """
+        if data_name is None:
+            data_name = self.gridlist
+        if file_name is None:
+            file_name = self.gridlist
+
+        if isinstance(data_name, str):
+            data_name = [data_name]
+        if isinstance(file_name, str):
+            file_name = [file_name]
+
+        for i, j in zip(data_name, file_name):
```

Comment:
change variable names.  `i` and `j` are usually indices and here they are confusingly used as variable/file names.